lib/Authen/SASL/Perl/XOAUTH2.pm

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) 2025 Aditya Garg <gargaditya08@live.com>
+# Copyright (c) 2025 Julian Swagemakers <julian@swagemakers.org>
+# All rights reserved.
+# This program is free software; you can redistribute it and/or
+# modify it under the same terms as Perl itself.
+
+package Authen::SASL::Perl::XOAUTH2;
+
+use strict;
+use vars qw(@ISA);
+
+@ISA     = qw(Authen::SASL::Perl);
+
+my %secflags = (
+    noanonymous => 1,
+);
+
+sub _order { 1 }
+
+sub _secflags {
+    shift;
+    scalar grep { $secflags{$_} } @_;
+}
+
+sub mechanism { 'XOAUTH2' }
+
+
+sub client_start {
+    my $self = shift;
+    $self->{stage} = 0;
+
+    # "user=" {User} "^Aauth=Bearer " {Access Token} "^A^A"
+    # https://developers.google.com/gmail/imap/xoauth2-protocol#initial_client_response
+    my $username = $self->_call('user');
+    my $token    = $self->_call('pass'); # OAuth 2.0 access token
+    my $auth_string = "user=$username\001auth=Bearer $token\001\001";
+    return $auth_string
+}
+
+sub client_step {
+    my ($self, $challenge) = @_;
+
+    my $stage = ++$self->{stage};
+
+    if ($stage == 1) {
+        # Send dummy request on authentication failure according to rfc7628.
+        # https://datatracker.ietf.org/doc/html/rfc7628#section-3.2.3
+        return "\001";
```

Comment:
In case of an error, the challenge is a base64 encoded JSON string. Would it be an idea to base64-decode that string, and then "unpack" the JSON into a Perl structure and set the error condition using `$self->set_error( $decoded_error )` ?

Reply:
Actually, the payload is a JSON string, the base64 encoding happens in the SMTP layer.

```perl
my $json = JSON::PP->new;
my $payload = $json->decode( $challenge );
$self->set_error( $payload );
```

Then, at the top, you need to:

```perl
use JSON::PP;
```

which is a Perl standard library, so it doesn't need to be added as a dependency *and* can be assumed to be available *everywhere*.

Reply:
Are you talking about `return $self->set_error("Invalid sequence");` here? Cause return "\001" seems to be simply following the spec.

Reply:
We can probably even simplify `client_step`, there are no other stages to be expected other than the failure case. What do you think of:

```perl
sub client_step {
    my ($self, $challenge) = @_;
    my $json = JSON::PP->new;
    my $payload = $json->decode( $challenge );
    $self->set_error( $payload );
    # Send dummy request on authentication failure according to rfc7628.
    # https://datatracker.ietf.org/doc/html/rfc7628#section-3.2.3
    return "\001";
}
```

Reply:
> We can probably even simplify `client_step`, there are no other stages to be expected other than the failure case. What do you think of:
> 
> ```perl
> sub client_step {
>     my ($self, $challenge) = @_;
>     my $json = JSON::PP->new;
>     my $payload = $json->decode( $challenge );
>     $self->set_error( $payload );
>     # Send dummy request on authentication failure according to rfc7628.
>     # https://datatracker.ietf.org/doc/html/rfc7628#section-3.2.3
>     return "\001";
> }
> ```

I was about to suggest something similar.

Reply:
But I am not sure how to test it

Reply:
Yes, if no further steps are ever expected other than the ERROR case, then this is a good implementation for `client_step()`. (would be nice if someone was able to trigger the error case before I accept the PR.)

Can this be tested by providing the wrong bearer token?

Reply:
> Can this be tested by providing the wrong bearer token?
Yes, you can do that, I'm tampering with the `auth_string` to make it invalid for manual testing. You could also revoke the client app from your Google/Microsoft account.

Reply:
> > Can this be tested by providing the wrong bearer token?
> 
> Yes, you can do that, I'm tampering with the `auth_string` to make it invalid for manual testing. You could also revoke the client app from your Google/Microsoft account.

SMTP generic error: Can't locate object method "new" via package "JSON::PP" (perhaps you forgot to load "JSON::PP"?) at /usr/share/perl5/Authen/SASL/Perl/XOAUTH2.pm line 41.

Reply:
`use JSON::PP;` is also there at top.

Reply:
I am passing wrong token btw

Reply:
Ok. Please push your latest state so I can see what the actual code looks like. The current state does not have JSON::PP at the top, so it's hard to see what else might be wrong.

Reply:
nvm, `libjson-pp-perl` was not installed by Ubuntu

Reply:
> Yes, if no further steps are ever expected other than the ERROR case, then this is a good implementation for `client_step()`. (would be nice if someone was able to trigger the error case before I accept the PR.)
> 
> Can this be tested by providing the wrong bearer token?

Error: HASH(0x58d0eb070630)
5.7.8 Username and Password not accepted. For more information, go to
5.7.8  https://support.google.com/mail/?p=BadCredentials 98e67ed59e1d1-309ef097cb7sm3466494a91.22 - gsmtp

Replicated

Reply:
> > Yes, if no further steps are ever expected other than the ERROR case, then this is a good implementation for `client_step()`. (would be nice if someone was able to trigger the error case before I accept the PR.)
> > Can this be tested by providing the wrong bearer token?
> 
> Error: HASH(0x58d0eb070630) 5.7.8 Username and Password not accepted. For more information, go to 5.7.8 https://support.google.com/mail/?p=BadCredentials 98e67ed59e1d1-309ef097cb7sm3466494a91.22 - gsmtp
> 
> Replicated

`print STDERR "Error: $payload\n";`
    
I had added this

Reply:
Ok. the output `Error: HASH(0x....)` means the payload was successfully converted from a JSON string to a Perl hash reference, which is the natural transformation from a javascript object to Perl. If you want to see what is in the hash, do:

```perl
use Data::Dumper;
print STDERR Dumper($payoad) . "\n";
```

instead.

Reply:
> Ok. the output `Error: HASH(0x....)` means the payload was successfully converted from a JSON string to a Perl hash reference, which is the natural transformation from a javascript object to Perl. If you want to see what is in the hash, do:
> 
> ```perl
> use Data::Dumper;
> print STDERR Dumper($payoad) . "\n";
> ```
> 
> instead.

$VAR1 = {
          'status' => 'invalid_request',
          'scope' => 'https://mail.google.com/'
        };

5.7.8 Username and Password not accepted. For more information, go to
5.7.8  https://support.google.com/mail/?p=BadCredentials d9443c01a7336-22db5216ba5sm32857675ad.222 - gsmtp

Reply:
> ```perl
> print STDERR Dumper($payoad) . "\n";
> ```

I quickly doublt checked, and with XOAUTH2, its a bit different

```
$VAR1 = {
          'schemes' => 'Bearer',
          'status' => '400',
          'scope' => 'https://mail.google.com/'
        };

5.7.8 Username and Password not accepted. For more information, go to
5.7.8  https://support.google.com/mail/?p=BadCredentials d9443c01a7336-22db5100aa7sm32888875ad.161 - gsmtp
```

Reply:
I'm not sure if it is significant, still I've Unresolved the conversation

Reply:
Cc @ehuelsmann

Reply:
@AdityaGarg8, thanks for drawing the additional attention. I saw your remark about them being different just before I merged. I'm fine with the difference. Not sure if we want/need to point this out in the documentation. Is this difference specified in the RFC?

Reply:
I haven't really found any docs as such

https://datatracker.ietf.org/doc/html/rfc7628#section-3.2.3